File: samples/book-app-project/book_app.py
import sys
from books import BookCollection


# 全域收藏實例
collection = BookCollection()


def show_books(books):
    """以使用者友好的格式顯示書籍。"""
    if not books:
        print("未找到書籍。")
        return

    print("\n您的圖書收藏：\n")

    for index, book in enumerate(books, start=1):
        status = "✓" if book.read else " "
        print(f"{index}. [{status}] {book.title} by {book.author} ({book.year})")

    print()


def handle_list():
    books = collection.list_books()
    show_books(books)


def handle_add():
    print("\n新增新書\n")

    title = input("標題：").strip()
    author = input("作者：").strip()
    year_str = input("年份：").strip()

    try:
        year = int(year_str) if year_str else 0
        collection.add_book(title, author, year)
        print("\n書籍新增成功。\n")
    except ValueError as e:
        print(f"\n錯誤：{e}\n")


def handle_remove():
    print("\n移除書籍\n")

    title = input("請輸入要移除的書籍標題：").strip()
    collection.remove_book(title)

    print("\n如果該書存在，則已移除。\n")


def handle_find():
    print("\n依作者尋找書籍\n")

    author = input("作者姓名：").strip()
    books = collection.find_by_author(author)

    show_books(books)


def show_help():
    print("""
圖書收藏協助工具

指令：
  list     - 顯示所有書籍
  add      - 新增一本書
  remove   - 依標題移除書籍
  find     - 依作者尋找書籍
  help     - 顯示此說明訊息
""")


def main():
    if len(sys.argv) < 2:
        show_help()
        return

    command = sys.argv[1].lower()

    if command == "list":
        handle_list()
    elif command == "add":
        handle_add()
    elif command == "remove":
        handle_remove()
    elif command == "find":
        handle_find()
    elif command == "help":
        show_help()
    else:
        print("未知指令。\n")
        show_help()


if __name__ == "__main__":
    main()

File: samples/book-app-project-js/book_app.js
const readline = require("readline");
const { BookCollection } = require("./books");

const collection = new BookCollection();

function showBooks(books) {
  if (!books || books.length === 0) {
    console.log("未找到書籍。");
    return;
  }

  console.log("\n您的圖書收藏：\n");

  books.forEach((book, index) => {
    const status = book.read ? "✓" : " ";
    console.log(`${index + 1}. [${status}] ${book.title} by ${book.author} (${book.year})`);
  });

  console.log();
}

function handleList() {
  const books = collection.listBooks();
  showBooks(books);
}

function prompt(question) {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });
  return new Promise((resolve) => {
    rl.question(question, (answer) => {
      rl.close();
      resolve(answer.trim());
    });
  });
}

async function handleAdd() {
  console.log("\n新增新書\n");

  const title = await prompt("標題：");
  const author = await prompt("作者：");
  const yearStr = await prompt("年份：");

  try {
    const year = yearStr ? parseInt(yearStr, 10) : 0;
    if (isNaN(year)) {
      throw new Error("年份必須是數字。");
    }
    collection.addBook(title, author, year);
    console.log("\n書籍新增成功。\n");
  } catch (err) {
    console.log(`\n錯誤：${err.message}\n`);
  }
}

function handleRemove() {
  return prompt("請輸入要移除的書籍標題：").then((title) => {
    console.log("\n移除書籍\n");
    collection.removeBook(title);
    console.log("\n如果該書存在，則已移除。\n");
  });
}

async function handleFind() {
  console.log("\n依作者尋找書籍\n");

  const author = await prompt("作者姓名：");
  const books = collection.findByAuthor(author);

  showBooks(books);
}

function showHelp() {
  console.log(`
圖書收藏協助工具

指令：
  list     - 顯示所有書籍
  add      - 新增一本書
  remove   - 依標題移除書籍
  find     - 依作者尋找書籍
  help     - 顯示此說明訊息
`);
}

async function main() {
  const args = process.argv.slice(2);

  if (args.length === 0) {
    showHelp();
    return;
  }

  const command = args[0].toLowerCase();

  switch (command) {
    case "list":
      handleList();
      break;
    case "add":
      await handleAdd();
      break;
    case "remove":
      await handleRemove();
      break;
    case "find":
      await handleFind();
      break;
    case "help":
      showHelp();
      break;
    default:
      console.log("未知指令。\n");
      showHelp();
      break;
  }
}

main();
